keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1004 @@
+  volumes:
++  - name: example-com-server-tls-trust-anchors
++    projected:
++      sources:
++      - trustAnchorSets:
++          signerName: example.com/server-tls
++          sets:
++          - singleFile:
++              set: default
++              path: ca_certificates.pem
+```
+
+When my client pod is scheduled onto a node, Kubelet writes the chosen trust
+anchor set into the container filesystem of my main container.  My main
+container doesn't start until `ca_certificates.pem` is present and non-empty.
+
+I ensure that my client application is capable of reading a set of trust anchors
+from a file containing PEM-wrapped DER-formatted X.509 certificates at the
+location I've specified
+(`/var/run/example-com-server-tls-trust-anchors/ca_certificates.pem`).
+
+I ensure that my client application notices changes to `ca_certificates.pem`
+(using inotify or periodic polling), and properly updates its TLS stack when
+that happens.
+
+#### TrustAnchorSets for Webhook Backends
+
+(To be added)
+
+### Risks and Mitigations
+
+Scalability: In the limit, TrustAnchorSets objects will be used by every pod in
+the cluster, which will require watches from all Kubelets..  When they are
+updated, workloads will need to receive the updates fairly quickly (within 5
+minutes across the whole cluster), to accommodate emergency rotation of trust
+anchors for a private CA.
+
+Security: Like CertificateSigningRequests, TrustAnchorSets will have a
+security-critical but "weird" integration with the RBAC system.  In particular,
+I expect that the controller that backs a signer will typically be granted
+access to update all TrustAnchorSets that have the corresponding
+`spec.signerName` field.
+
+Security: Should Kubelet attempt to enforce which workloads are allowed to mount
+particular TrustAnchorSets using a trustAnchorSets volume?  The straightforward
+design would have all TrustAnchorSet reads happen using a `system:node:xyz`
+identity, which would allow any workload to mount any TrustAnchorSets object.
+
+Security: Should individual trust anchor set entries designate an OSCP endpoint
+to check for certificate revociation?  Or should we require the URL to be
+embedded in the issued certificates?
+
+Reliability: How can we safely canary a change to a TrustAnchorSets object?
+
+Usability: Is it always necessary to provide a value for spec.signerName?
+
+## Design Details
+
+### TrustAnchorSets Object
+
+The TrustAnchorSets object allows signers to publish their trust anchors for
+consumption by workloads, whether directly, via Kubelet, or otherwise.
+
+Every signer corresponds to zero, one, or more TrustAnchorSets objects.  The
+name of the TrustAnchorSets objects are arbitrary; the association is managed
+entirely via the `spec.signerName` field, which is immutable.
+
+Since a given signer could have multiple TrustAnchorSets objects, clients
+looking for a particular named set X must take the union of all `spec.sets`
+entries where `spec.sets[i].name == X`.
```

Comment:
Shouldn't the kubelet do this for you?

Reply:
In this case, "Clients" is "APIServer Clients", i.e. the Kubelet would be the client this refers to.

Reply:
This section has been removed now that we use a fixed mapping between signer names and TrustAnchorSets object names.